cache/test_cache.py

```diff
@@ -68,6 +68,11 @@ def test_cache_fulfill_prefix(self):
             self.mixed_config(),
             self.chain(cache_alowed=True, uri='/static/content.html'))
 
+    def test_cache_deception_attack(self):
+        self.generic_test_routine(
+            self.mixed_config(),
+            self.chain(cache_alowed=False, uri='/home.php/picts/bear.jpg'))
```

Comment:
Yes, now the test fails, and the request is cached, because it has the suffix `jpg`. But some web servers interpret this url as a call to home.php, and caching home.php can lead to data leaks. See the article from the issue

Reply:
Yeah, I've read the article.

The vulnerability needs (1) a backend server to explicitly state that its response should not be cached, and (2) a caching proxy to cache anyway. Here you have the rules to cache every response which URI ends on `.jpg`, but have no cache prohibiting directive from the backend server.

As far as I understand, the framework named "deproxy" (local customized version) is used to emulate a backend. Combination of the client request, the request forwarded to the backend, the backend response, and the response forwarded to the client, is called "chain". By calling already existing functions you are generating a chain where the backend response have no explicit caching directives. Thus Tempesta caches response based on URI suffix. That's why the test fails.

This test is to test what happens if a backend server returns `Cache-Control: no-store` while Tempesta is configured to cache a response based on an URL suffix. To test that, one needs the backend server, or in this case deproxy framework, to generate a response with `Cache-Control: no-store`.

Reply:
Oh, by the way. I've just merged typo correcting PR, which among other typos changes `cache_alowed` to `cache_allowed`. Please rebase your branch and change to `cache_allowed` too.

Reply:
I agree with @i-rinat : the test isn't consistent and doesn't close the original issue. The article says that proxy server follows mach rules for populating the cache (`cache_fulfill` directive in our case), but doesn't honour cache management headers sent by origin server (e.g. `Cache-Control: no-cache`). Tempesta is not vulnerable for the attack, since we check headers before put response into cache:
https://github.com/tempesta-tech/tempesta/blob/c7dc1fd5873be78ea141315cb23828561ddb8553/tempesta_fw/cache.c#L1343

But this test doesn't check the case, since response generated from "origin" server doesn't contain `Cache-Control: no-cache` or similar directives. According to RFC, the response can be cached by intermediaries in this case, and Tempesta actually does it. You can use the 5-th level of verbose (`-vvvvv`) when running the tests to see all HTTP messages between Tempesta and tests framework.

Please consult to `tfw_cache_employ_resp()` function in Tempesta code to clarify, which responses can't be cached, there are more cases in addition to `Cache-Control: no-cache`. All of them must be tested. Moreover, I  would like to ask you to look at [Section 3 of RFC 7234](https://tools.ietf.org/html/rfc7234#section-3) and check that responses are cached only if it's allowed by RFC. 

By the way, if Tempesta violates RFC (or declared behaviour) and the tests fails, it's a subject to create a new issue in Tempesta repo.